bundles/framework/hierarchical-layerlist/Flyout.js

```diff
@@ -0,0 +1,749 @@
+        /* PRIVATE METHODS
+        *******************************************************************************************************************************/
+        /**
+         * Bind extender service listeners
+         * @method  _bindExtenderServiceListeners
+         * @private
+         */
+        _bindExtenderServiceListeners: function() {
+            var me = this;
+
+            me.service.on('option.added', function() {
+                me.populateLayers();
+            });
+
+            me.service.on('admin.changed', function() {
+                me.populateLayers();
+            });
+
+            me.service.on('admin-layer', function(data) {
+                var mode = data.mode;
+
+                if (mode === 'delete') {
+                    me.mapLayerService.removeLayer(data.id);
+                } else if (mode === 'add') {
+                    var layer = me.instance.sandbox.findMapLayerFromAllAvailable(data.layerData.id);
+                    me.mapLayerService.addLayer(layer);
+                    me.mapLayerService.updateLayersInGroups(data.layerData.id, data.layerData, false, true);
+                } else {
+                    me.mapLayerService.updateLayer(data.layerData.id, data.layerData);
+
+                    // also update breadcrump information
+                    me._updateBreadcrumbGroups(data.layerData);
+                }
+
+
+                me.populateLayers();
+            });
+
+            me.service.on('jstree-order-changed', function(data) {
+                // order not changed, need reload tree
+                if (data.ajax && data.success === false) {
+                    me.populateLayers();
+                }
+            });
+
+
+            // group added
+            me.service.on('group-added', function(data) {
+                // update service groups
+                me.mapLayerService.updateLayerGroups(data);
+                me.populateLayers();
+            });
+
+            // group deleted
+            me.service.on('group-deleted', function(data) {
+                me.mapLayerService.deleteLayerGroup(data.id, data.parentId);
+                me.populateLayers();
+            });
+        },
+        /**
+         * Update breadcrumbs
+         * @method  _updateBreadcrumbGroups
+         * @param   {Object}                layerData layer data
+         * @private
+         */
+        _updateBreadcrumbGroups: function(layerData) {
+            var me = this;
+            if (me.instance._selectedLayerGroupId[layerData.id]) {
+                var currentGroupId = me.instance._selectedLayerGroupId[layerData.id];
+                var founded = jQuery.grep(layerData.groups, function(group) {
+                    return group.id === currentGroupId;
+                });
+
+                if (founded.length === 0) {
+                    // update groups to first
+                    me.instance._selectedLayerGroupId[layerData.id] = layerData.groups[0].id;
+                }
```

Comment:
You can write this with array.find() (it's polyfilled in core now):

    _updateBreadcrumbGroups: function(layerData) {
        var me = this;
        var currentGroupId = me.instance._selectedLayerGroupId[layerData.id];
        if (!currentGroupId) {
            // requested group not found - ignoring
            return;
        }
        var group = layerData.groups.find(function(group) {
            return group.id === currentGroupId;
        });
        
        if(group) {
            // update groups to first
            me.instance._selectedLayerGroupId[layerData.id] = layerData.groups[0].id;
        }
    }

Though looking at the code it seems kinda fishy. You check if the layer is included in me.instance._selectedLayerGroupId (dictionary of layer id -> group id). If it is you check if the layer has the group that was linked to the layer id included as one of the layers groups. If it is you overwrite the layer to group mapping value with the id of first group for the layer? Does this have something to do with breadcrumbs? The dictionary value for the layer is always the first of the layers groups, but it's tracked with another variable in instance? Maybe naming could be improved so this doesn't feel so disconnected from it's name?